File: classes04/python/delaunayAnimation.py
"""
delaunayAnimation.py
==========
@brief 使用动画展示Delaunay三角剖分和Voronoi的划分的过程

@file delaunayAnimation.cpp
@author Shawn Zhang
@date 2018-09-09
"""
import cv2
import dlib
import numpy as np
import random

def rectContains(rect, point):
    """
    检查点是否在矩形内部
    参数
    --------------
    rect:矩形 [x,y,width,height]
    point:点 [x,y]
    返回值
    --------------
    True -> 点在矩形内部
    False -> 点在矩形外
    """
    pass

    if point[0]<rect[0]:
        return False
    elif point[1]<rect[1]:
        return False
    elif point[0]>(rect[2]+rect[0]):
        return False
    elif point[1]>(rect[3]+rect[1]):
        return False
    return True

def faceLandmarksDetect(im):
    """
    获取脸部的关键特征点
    参数
    ---------
    im:需要检测的图像
    返回值
    ---------
    points:检测到的关键点
    """
    pass

    PREDICTOR_PATH="../../common/resources/zxm_shape_predictor_70_face_landmarks.dat"
    faceDetector = dlib.get_frontal_face_detector()
    shapePredictor = dlib.shape_predictor(PREDICTOR_PATH)

    faceRects=faceDetector(im,0)
    points=[]
    if len(faceRects)>0:
        newRect=dlib.rectangle(int(faceRects[0].left()),int(faceRects[0].top()),int(faceRects[0].right()),int(faceRects[0].bottom()))
        landmarks=shapePredictor(im,newRect)
        for p in landmarks.parts():
            points.append((p.x, p.y))
    print("检测到的脸部关键点为:")
    for index, value in enumerate(points):
        print("{}:{}".format(index, value))
    
    return points

def drawPoint(img, p, color):
    """
    绘制点
     参数
     -----------
    img:绘制的图像;
    p: 点坐标;
    color:绘制用的颜色
    """
    pass

    cv2.circle(img, p, 2, color)

def drawDelaunay(img, subdiv, delaunayColor):
    """
    绘制Delaunay剖分三角形
    参数
    ---------
    img:绘制的图像;
    subdiv:剖分器;
    delaunayColor:绘制用的颜色
    """
    pass

    triangleList = subdiv.getTriangleList()
    size = img.shape
    r=(0,0,size[1],size[0])

    for t in triangleList:
        pt1 = (t[0],t[1])
        pt2 = (t[2],t[3])
        pt3 = (t[4],t[5])

        if rectContains(r, pt1) and rectContains(r, pt2) and rectContains(r, pt3):
            cv2.line(img, pt1, pt2, delaunayColor)
            cv2.line(img, pt2, pt3, delaunayColor)
            cv2.line(img, pt3, pt1, delaunayColor)

def drawVoronoi(img, subdiv):
    """
    绘制Voronoi图
    参数
    ------------
    img:绘制的画板;
    subdiv:剖分器
    """
    pass

    (facets, centers)=subdiv.getVoronoiFacetList([])

    for i in range(0, len(facets)):
        ifaceArr = []
        for f in facets[i]:
            ifaceArr.append(f)
        
        ifacet = np.array(ifaceArr, np.int)

        # 生成随机颜色
        color = (random.randint(0,255),random.randint(0,255),random.randint(0,255))

        # 绘制面
        cv2.fillConvexPoly(img, ifacet, color, cv2.LINE_AA, 0)

        # 绘制边界
        ifacets=np.array([ifacet])
        cv2.polylines(img, ifacets, True, (0,0,0), 1, cv2.LINE_AA, 0)

        # 绘制中心
        cv2.circle(img, (centers[i][0], centers[i][1]), 2, (0,0,0), -1, cv2.LINE_AA, 0)

# 入口
if __name__ == "__main__":
    win="Delaunay Animation"
    delaunayColor = (255, 255, 255)
    pointsColor = (0, 0, 255)

    img = cv2.imread("../data/images/smiling-man.jpg")

    rect=(0, 0, img.shape[1], img.shape[0])
    subdiv = cv2.Subdiv2D(rect)

    # 获取脸部的关键点
    points = faceLandmarksDetect(img)

    imgVoronoi = np.zeros(img.shape, dtype=img.dtype)
    imgDelaunay = img.copy()

    # 绘制
    plotPoints = []
    for p in points:
        drawPoint(img, p, pointsColor)
        subdiv.insert(p)
        plotPoints.append(p)

        imgDelaunay = img.copy()

        drawDelaunay(imgDelaunay, subdiv, delaunayColor)
        drawVoronoi(imgVoronoi,subdiv)

        imDisplay = np.hstack([imgDelaunay, imgVoronoi])
        cv2.imshow(win, imDisplay)
        cv2.waitKey(400)
    

    cv2.waitKey(0)
    cv2.destoryAllWindows()
    

File: classes04/cpp/delaunayAnimation.cpp
/**
 * @brief 使用动画展示Delaunay三角剖分和Voronoi的划分的过程
 * 
 * @file delaunayAnimation.cpp
 * @author Shawn Zhang
 * @date 2018-09-09
 */
#include <dlib/opencv.h>
#include <dlib/image_processing.h>
#include <dlib/image_processing/frontal_face_detector.h>
#include <opencv2/opencv.hpp>

using namespace cv;
using namespace std;
using namespace dlib;

/**
 * @brief 在图像上绘制特征点
 * 
 * @param img 绘制的图像
 * @param fp 特征点的坐标
 * @param color 绘制使用的颜色
 */
static void drawPoint(Mat &img, Point2f fp, Scalar color)
{
    cv::circle(img, fp, 2, color);
}

/**
 * @brief 在图像上绘制Delaunay三角形
 * 
 * @param img 要绘制的图形
 * @param subdiv 划分类
 * @param delaunayColor 绘制使用的颜色 
 */
static void drawDelaunay(Mat &img, Subdiv2D &subdiv, Scalar delaunayColor)
{
    
    // 获取三角形列表
    std::vector<cv::Vec6f> triangleList;
    subdiv.getTriangleList(triangleList);

    std::vector<Point> vertices(3);
    Size size = img.size();
    Rect rect(0, 0, size.width, size.height);

    for (size_t i = 0; i < triangleList.size(); i++)
    {
        Vec6f t = triangleList[i];

        // 转三角形为列表
        vertices[0] = Point(cvRound(t[0]), cvRound(t[1]));
        vertices[1] = Point(cvRound(t[2]), cvRound(t[3]));
        vertices[2] = Point(cvRound(t[4]), cvRound(t[5]));


        // 绘制三角形
        if (rect.contains(vertices[0]) && rect.contains(vertices[1]) && rect.contains(vertices[2]))
        {
            cv::line(img, vertices[0], vertices[1], delaunayColor);
            cv::line(img, vertices[1], vertices[2], delaunayColor);
            cv::line(img, vertices[2], vertices[0], delaunayColor);
        }
    }
}

/**
 * @brief 绘制沃罗诺伊图
 * 
 * @param img 要绘制的图像
 * @param subdiv 划分器
 */
static void drawVoronoi(Mat &img, Subdiv2D &subdiv)
{
    // voronoi面片列表
    std::vector<std::vector<Point2f>> facets;

    // voronoi面片的中心点
    std::vector<Point2f> centers;

    // 获取voronoi图的面片列表和中心点
    subdiv.getVoronoiFacetList(std::vector<int>(), facets, centers);

    // 使用fillConvexPoly方法绘制面片
    std::vector<Point> ifacet;
    // 使用polylines方法绘制面片的边缘
    std::vector<std::vector<Point>> ifaces(1);
    for(size_t i=0; i<facets.size(); i++)
    {
        ifacet.resize(facets[i].size());
        for(size_t k=0; k<facets[i].size(); k++)
        {
            ifacet[k]=facets[i][k];
        }

        // 生成随机颜色
        Scalar color;
        color[0] = std::rand() & 255; // 这是一个小诀窍，因为255的二进制位11111111，所以去与计算，最大值只能是255
        color[1] = std::rand() & 255;
        color[2] = std::rand() & 255;

        // 填充面片
        cv::fillConvexPoly(img, ifacet, color);

        // 绘制面片的边界
        ifaces[0] = ifacet;
        cv::polylines(img, ifaces, true, Scalar());

        // 绘制中心点
        cv::circle(img, centers[i], 3, Scalar());
    }
}

/**
 * @brief Get the Face Landmarks object
 * 
 * @param img 
 * @return std::vector<Point> 
 */
std::vector<Point2f> getFaceLandmarks(Mat& img)
{
    // 面部检测器
    dlib::frontal_face_detector faceDetector = get_frontal_face_detector();
    // 形状检测器
    dlib::shape_predictor landmarkPredictor;

     // 加载面部关键特征点模型
    deserialize("../../common/resources/zxm_shape_predictor_70_face_landmarks.dat") >> landmarkPredictor;

    // 将OpenCV的图像格式转换为dlib的图像格式
    cv_image<bgr_pixel> dlibImg(img);
    std::vector<dlib::rectangle> faceRects=faceDetector(dlibImg);
    
    dlib::full_object_detection landmarks= landmarkPredictor(dlibImg, faceRects[0]);

    std::vector<Point2f> points;
    for(int i=0; i<landmarks.num_parts();i++)
    {
        points.push_back(Point2f(landmarks.part(i).x(), landmarks.part(i).y()));
        cout<<"特征点"<<i<<":"<<points[i]<<endl;
    }

    

    return points;

}

int main(int argc, char const *argv[])
{
    string win="Delaunay Animation";
    Scalar delaunayColor(255, 255, 255), pointsColor(0, 0, 255);
    Mat img = imread("../data/images/smiling-man.jpg");

    // 初始化一个划分器
    Rect rect(1,1,img.size().width, img.size().height);
    Subdiv2D subdiv(rect);

    // 获取关键点点
    std::vector<Point2f> points = getFaceLandmarks(img);

    // 显示三角剖分的图像
    Mat imgDelaunay;
    // 显示Voronoi图的图像
    Mat imgVoronoi = Mat::zeros(img.rows, img.cols, CV_8UC3);

    // 最终显示的图像
    Mat imgDisplay;

    // 在划分器中插入点进行划分
    for(std::vector<Point2f>::iterator it = points.begin();it != points.end(); it++)
    {
        subdiv.insert(*it);

        imgDelaunay = img.clone();
        imgVoronoi = cv::Scalar(0,0,0);

        // 绘制delaunay
        drawDelaunay(imgDelaunay, subdiv, delaunayColor);

        // 绘制点
        for(std::vector<Point2f>::iterator ptr=points.begin(); ptr != points.end(); ptr++ )
        {
            drawPoint(imgDelaunay, *ptr, pointsColor);
        }

        // 绘制Voroni
        drawVoronoi(imgVoronoi, subdiv);
        cv::hconcat(imgDelaunay, imgVoronoi, imgDisplay);
        imshow(win, imgDisplay);
        waitKey(400);
    }


    waitKey(0);

    return 0;
}

